compiler/next/lib/parsing/ParserContextImpl.h

```diff
@@ -121,6 +124,124 @@ owned<Expression> ParserContext::consumeVarDeclLinkageName(void) {
   return toOwned(ret);
 }
 
+owned<Attributes> ParserContext::buildAttributes(YYLTYPE locationOfDecl) {
+  numAttributesBuilt += 1;
+
+  // There may be nothing to return.
+  if (!hasAttributeParts) {
+    return nullptr;
+  }
+
+  // Have to copy construct.
+  auto pragmaCopy = attributeParts.pragmas
+      ? *(attributeParts.pragmas)
+      : std::set<PragmaTag>();
+
+  auto node = Attributes::build(builder, convertLocation(locationOfDecl),
+                                std::move(pragmaCopy),
+                                attributeParts.isDeprecated,
+                                attributeParts.deprecationMessage);
+  return node;
+}
+
+PODUniqueString ParserContext::notePragma(YYLTYPE loc,
+                                          Expression* pragmaStr) {
+  auto ret = PODUniqueString::build();
+
+  // Extract the string literal and convert it into a pragma flag.
+  if (auto strLit = pragmaStr->toStringLiteral()) {
+    ret = PODUniqueString::build(context(), strLit->str().c_str());
+    auto tag = pragmaNameToTag(ret.c_str());
+
+    if (tag == PRAGMA_UNKNOWN) {
+      std::string msg;
+      msg += "Unrecognized compiler pragma: ";
+      msg += ret.c_str();
+      noteError(loc, msg.c_str());
+    } else {
+
+      // Initialize the pragma flags if needed.
+      auto& pragmas = attributeParts.pragmas;
+      if (pragmas == nullptr) {
+        pragmas = new std::set<PragmaTag>();
+      }
+
+      hasAttributeParts = true;
+      pragmas->insert(tag);
+    }
+  }
+
+  // Make sure to clean up the UAST node since it will be discarded.
+  delete pragmaStr;
+
+  return ret;
+}
+
+void ParserContext::noteDeprecation(YYLTYPE loc, Expression* messageStr) {
+  if (!hasAttributeParts) {
+    hasAttributeParts = true;
+  } else {
+    assert(!attributeParts.isDeprecated);
+    assert(attributeParts.deprecationMessage.isEmpty());
```

Comment:
Shouldn't trying to put two `deprecated` on the same symbol be a user-facing error? (You could do that now or add a TODO comment)